daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala

```diff
@@ -734,17 +734,21 @@ private[lf] final class Compiler(
       },
     )
 
-  @inline
-  private[this] def compileELet(elet: ELet) =
-    withEnv { _ =>
-      elet match {
-        case ELet(Binding(optBinder, _, bound), body) =>
-          let(withOptLabel(optBinder, compile(bound))) { boundPos =>
-            optBinder.foreach(addExprVar(_, boundPos))
-            compile(body)
-          }
-      }
+  // Compile nested lets using constant stack.
+  @tailrec
+  private[this] def compileELet(
+      eLet0: ELet,
+      bindings: mutable.Builder[SExpr, List[SExpr]] = List.newBuilder,
+  ): SELet = {
+    val binding = eLet0.binding
+    bindings += withOptLabel(binding.binder, compile(binding.bound))
+    val boundPos = nextPosition()
+    binding.binder.foreach(addExprVar(_, boundPos))
+    eLet0.body match {
+      case eLet1: ELet => compileELet(eLet1, bindings)
```

Comment:
you’re still merging nested lets here which is only valid if we can have references between bindings in a let so my previous comment still holds

Reply:
> do we need to worry about the inner let having references to variables bound in outer let? I forgot if you can have references in multi-lets to previous bindings in daml-lf.

We do not have to worry. In Speedy (like in the proto),  the  `Let` construction with multiple binding is a compressed form of nested let, each bound have access to the variable of previous bound. 

In any case you can see the function `let` using in the original implementation was already merging the `SELet`.

Reply:
great thanks for confirming